In [1]:
import json
import pandas as pd

In [2]:
def _partial_acc(obs):
    """
        inner function used to calculate a weighted accuracy based on how many human
        raters' annotations the predicted value matches.  Applied to a single Series.
        
        Args:
            obs = pandas.Series object containing 'predicted_answer' and 'annotations' columns
    """
    prediction = obs['predicted_answer']
    annotations = obs['annotations']
    matches = 0
    for a in annotations:
        if prediction.strip().lower() == a.strip().lower():
            matches += 1
    return min(1, matches/3)

## Download appropriate results file from Google Storage bucket.

Base Model (Top 1000 classification):
- `gs://mids-w266-mw/test2015_results_san_expt0_2018-12-06-05:05:51.json`

Best Enhanced Model (Top 1000 classification):
- `gs://mids-w266-mw/mlflow/52/b25acc49a4b04ac48d0d9e5be5a4f020/artifacts/test2015_results_mrr_san_expt28_2018-12-07-19:57:00.json`

Update `json_results_path` variable in the next cell:

### 1. Best "enhanced" model

In [3]:
# Best Enhanced Model (Top 1000 classification)
json_results_path = '/home/mwinton/report_results/test2015_results_mrr_san_expt28_2018-12-07-19:57:00.json'

with open(json_results_path) as f:
    resultsj = json.load(f)
    
df = pd.DataFrame(resultsj)
df['correct'] = (df['answer_str'].str.strip().str.lower() == df['predicted_answer'].str.strip().str.lower()).astype(int)
df['partial'] = df.apply(_partial_acc, axis=1)

In [4]:
df.describe()

answer_id       image_id  one_hot_index   question_id       correct  \
count  5.221300e+04   52213.000000   52213.000000  5.221300e+04  52213.000000   
mean   2.912509e+07  291250.822515      84.372513  2.912509e+06      0.535518   
std    1.682681e+07  168268.075268     181.851758  1.682681e+06      0.498742   
min    4.200000e+03      42.000000       1.000000  4.200000e+02      0.000000   
25%    1.467572e+07  146757.000000       1.000000  1.467572e+06      0.000000   
50%    2.918340e+07  291834.000000       5.000000  2.918340e+06      1.000000   
75%    4.361721e+07  436172.000000      58.000000  4.361721e+06      1.000000   
max    5.819132e+07  581913.000000    1000.000000  5.819132e+06      1.000000   

            partial  
count  52213.000000  
mean       0.615990  
std        0.465337  
min        0.000000  
25%        0.000000  
50%        1.000000  
75%        1.000000  
max        1.000000

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52213 entries, 0 to 52212
Data columns (total 13 columns):
annotations         52213 non-null object
answer_id           52213 non-null int64
answer_str          52213 non-null object
answer_type         52213 non-null object
complement_id       0 non-null object
image_id            52213 non-null int64
one_hot_index       52213 non-null int64
predicted_answer    52213 non-null object
question_id         52213 non-null int64
question_str        52213 non-null object
question_type       52213 non-null object
correct             52213 non-null int64
partial             52213 non-null float64
dtypes: float64(1), int64(5), object(7)
memory usage: 5.2+ MB


In [6]:
df.head()

annotations  answer_id answer_str  \
0           [16, 16, 16, 16, 16, 16, 16, 16, 16, 16]    9786500         16   
1  [yes, yes, yes, yes, yes, yes, yes, yes, yes, ...    9786510        yes   
2  [container, frisbee golf, frisbee golf goal, f...    9786520    frisbee   
3      [garbage, no, no, no, no, no, no, no, no, no]   57484500         no   
4  [yes, yes, yes, yes, yes, yes, yes, yes, yes, ...   57484510        yes   

  answer_type complement_id  image_id  one_hot_index predicted_answer  \
0      number          None     97865            241     fire hydrant   
1      yes/no          None     97865              1               no   
2       other          None     97865             24     fire hydrant   
3      yes/no          None    574845              2               no   
4      yes/no          None    574845              1              yes   

   question_id                     question_str question_type  correct  \
0       978650                     what # is it          what        0   
1       978651             is there people here      is there        0   
2       978652  what is the object on the right   what is the        0   
3      5748450                   is this inside       is this        1   
4      5748451         could someone sleep here         could        1   

   partial  
0      0.0  
1      0.0  
2      0.0  
3      1.0  
4      1.0

In [7]:
# count by answer type
df.groupby(['answer_type'])['correct'].count().sort_values(ascending=False)

answer_type
yes/no    22792
other     22686
number     6735
Name: correct, dtype: int64

In [8]:
pd.set_option('display.max_rows', 75)

In [9]:
# count by question type
df.groupby(['question_type'])['correct'].count().sort_values(ascending=False)

question_type
how many                    5478
is the                      4871
what color is the           4019
what                        2952
what is the                 2342
is this                     2180
none of the above           2038
is this a                   2034
are the                     1577
what is                     1294
is there a                  1245
what kind of                1178
is it                       1020
does the                     924
is                           900
is there                     891
what color are the           872
what type of                 803
are these                    755
are there                    709
is the man                   662
what are the                 645
where is the                 640
are                          636
does this                    607
what is the man              603
which                        557
how many people are          512
do                           445
what is on the               

In [10]:
# accuracy overall
acc = df['correct'].mean()
partial_acc = df['partial'].mean()
print('Accuracy = {:.3f}. Partial Accuracy = {:.3f}.'.format(acc, partial_acc))

Accuracy = 0.536. Partial Accuracy = 0.616.


In [11]:
# accuracy by question type
acc_by_qtype = df.groupby(['question_type'])['correct','partial'] \
    .mean() \
    .sort_values(['correct'], ascending=False)
acc_by_qtype

correct   partial
question_type                               
what room is              0.864407  0.878531
what sport is             0.858934  0.881923
is there a                0.852209  0.909505
could                     0.813953  0.905039
are there                 0.739069  0.801128
do you                    0.734375  0.833333
can you                   0.734177  0.801688
is there                  0.728395  0.808081
was                       0.718876  0.792503
is the woman              0.715302  0.791222
is                        0.684444  0.784444
is it                     0.684314  0.764706
does the                  0.682900  0.774892
has                       0.680769  0.789744
is this an                0.678414  0.795888
what animal is            0.674107  0.690476
is the person             0.673913  0.760870
is the man                0.667674  0.770393
does this                 0.667216  0.775398
is this a                 0.658800  0.771878
are the                   0.655041  0.761995
is this person            0.654206  0.777259
are they                  0.650407  0.751581
is the                    0.647506  0.755218
is this                   0.645413  0.760245
is he                     0.641304  0.760870
are these                 0.638411  0.772185
are                       0.625786  0.755765
do                        0.624719  0.737079
are there any             0.619178  0.751598
is that a                 0.609195  0.722222
what brand                0.594595  0.648649
what type of              0.574097  0.636779
what is this              0.572626  0.618250
what is the man           0.570481  0.622996
what kind of              0.556027  0.616299
what is the person        0.554502  0.617694
what are                  0.528239  0.570321
none of the above         0.522571  0.630684
what is the name          0.500000  0.500000
what                      0.490515  0.539747
what are the              0.469767  0.520413
what is the woman         0.467662  0.502488
what does the             0.460000  0.496667
what is the color of the  0.459227  0.510730
what time                 0.455090  0.624750
what is the               0.450470  0.493595
what color are the        0.449541  0.505734
what color is the         0.434685  0.489342
which                     0.421903  0.488929
what is on the            0.391924  0.431512
who is                    0.373391  0.422031
what color is             0.371758  0.447646
what is                   0.363215  0.418083
where are the             0.358650  0.431786
what is in the            0.333333  0.387636
what color                0.303116  0.340888
why is the                0.298701  0.359307
how many                  0.292808  0.384143
how many people are       0.289062  0.369792
why                       0.287425  0.307385
where is the              0.282813  0.363542
how                       0.281915  0.322695
how many people are in    0.243243  0.312312
what number is            0.050847  0.050847

In [12]:
# accuracy by answer type
acc_by_anstype = df.groupby(['answer_type'])['correct','partial'] \
    .mean() \
    .sort_values(['correct'], ascending=False)
acc_by_anstype

correct   partial
answer_type                    
yes/no       0.678703  0.784208
other        0.468086  0.521717
number       0.278099  0.364266

In [13]:
acc_by_anstype.to_dict('index')

{'yes/no': {'correct': 0.6787030537030537, 'partial': 0.7842079092079145},
 'other': {'correct': 0.46808604425636957, 'partial': 0.5217167709894512},
 'number': {'correct': 0.2780994803266518, 'partial': 0.36426627072506856}}

### 1a. Yes/No Answer Type

In [14]:
# accuracy by question type
df[df.answer_type=='yes/no'].groupby(['question_type'])['correct', 'partial'].mean().sort_values(['correct'])

correct   partial
question_type                        
what is the        0.000000  0.000000
what               0.250000  0.250000
why                0.500000  0.500000
are there any      0.619178  0.751598
do                 0.621005  0.735160
none of the above  0.621644  0.769855
are                0.631746  0.761905
is that a          0.638710  0.763441
are these          0.642857  0.777015
is he              0.645756  0.767528
is this            0.649828  0.768024
is the             0.658422  0.767305
are the            0.661290  0.771953
is this a          0.663726  0.777201
are they           0.665730  0.765918
is this person     0.668317  0.798680
does this          0.671717  0.781145
is the man         0.674566  0.779358
is it              0.678986  0.763464
has                0.680769  0.789744
is this an         0.685446  0.810642
does the           0.687220  0.778401
is                 0.691860  0.792248
is the person      0.695876  0.786942
was                0.713656  0.792952
is the woman       0.727941  0.805147
do you             0.731579  0.831579
is there           0.731597  0.809362
are there          0.747839  0.809798
can you            0.750000  0.813218
could              0.813953  0.905039
is there a         0.852776  0.909895

In [15]:
# examples of 0% accuracy for yes/no answer type - 7 data points
df[(df.answer_type=='yes/no') & \
   ((df.question_type=='why') | (df.question_type=='what') | (df.question_type=='what is the'))]

annotations  answer_id  \
7754   [yes, yes, yes, company logo, yes, yes, yes, y...   22455400   
19452           [no, no, no, no, no, no, no, no, no, no]    8474900   
20140  [make feeding easier, yes, yes, yes, yes, view...   18663720   
23660       [no, no, no, no, no, yes, no, yes, yes, yes]    8276500   
44443  [no, emergency, no, no, ve, no, no, can't see,...    3987110   
44703  [yes, 2 men skateboarding, old picture, color,...   27829010   
50225  [yes, his preference, yes, yes, yes, because h...    9924220   

      answer_str answer_type complement_id  image_id  one_hot_index  \
7754         yes      yes/no          None    224554              1   
19452         no      yes/no          None     84749              2   
20140        yes      yes/no          None    186637              1   
23660         no      yes/no          None     82765              2   
44443         no      yes/no          None     39871              2   
44703        yes      yes/no          None    278290              1   
50225        yes      yes/no          None     99242              1   

      predicted_answer  question_id  \
7754            tracks      2245540   
19452               no       847490   
20140              yes      1866372   
23660           laptop       827650   
44443           subway       398711   
44703  black and white      2782901   
50225             cold       992422   

                                          question_str question_type  correct  \
7754                   what there a crown on the train          what        0   
19452                  what this picture taken indoors          what        1   
20140  why is there a wooden platform behind the fence           why        1   
23660           what someone using the computer in bed          what        0   
44443                     what word is show on the bus          what        0   
44703              what is the picture white and black   what is the        0   
50225                    why does the man have a beard           why        0   

       partial  
7754       0.0  
19452      1.0  
20140      1.0  
23660      0.0  
44443      0.0  
44703      0.0  
50225      0.0

In [16]:
# examples of <60% accuracy for yes/no answer type with correct answers - 914 data points
df[(df.answer_type=='yes/no') & \
   ((df.question_type=='are there any') | (df.question_type=='none of the above')) & \
   (df.correct==1)].tail(10)

annotations  answer_id  \
51801  [yes, yes, yes, yes, yes, yes, yes, yes, yes, ...   41001900   
51845         [no, yes, no, no, no, no, no, yes, no, no]   53550600   
51852            [no, no, no, no, no, no, no, no, no, 0]   21420400   
51888           [no, no, no, no, no, no, no, no, no, no]   24424600   
51911  [yes, maybe, yes, yes, yes, yes, yes, yes, yes...   52385410   
51979  [yes, yes, yes, yes, yes, yes, yes, yes, yes, ...   28985510   
52020  [yes, yes, yes, yes, yes, yes, yes, yes, yes, ...   40771720   
52055           [no, no, no, no, no, no, no, no, no, no]   25803610   
52152     [no, yes, yes, no, no, yes, yes, yes, yes, no]    6783220   
52168  [yes, no, yes, yes, yes, yes, yes, yes, yes, yes]   32166500   

      answer_str answer_type complement_id  image_id  one_hot_index  \
51801        yes      yes/no          None    410019              1   
51845         no      yes/no          None    535506              2   
51852         no      yes/no          None    214204              2   
51888         no      yes/no          None    244246              2   
51911        yes      yes/no          None    523854              1   
51979        yes      yes/no          None    289855              1   
52020        yes      yes/no          None    407717              1   
52055         no      yes/no          None    258036              2   
52152        yes      yes/no          None     67832              1   
52168        yes      yes/no          None    321665              1   

      predicted_answer  question_id  \
51801              yes      4100190   
51845               no      5355060   
51852               no      2142040   
51888               no      2442460   
51911              yes      5238541   
51979              yes      2898551   
52020              yes      4077172   
52055               no      2580361   
52152              yes       678322   
52168              yes      3216650   

                                            question_str      question_type  \
51801  should this sport be done away from house stru...  none of the above   
51845                                does she look happy  none of the above   
51852              are there any towels in this bathroom      are there any   
51888                    are there any clouds in the sky      are there any   
51911                            can someone eat outside  none of the above   
51979           are there any palm trees in this picture      are there any   
52020                   would a vegetarian eat this food  none of the above   
52055                       does she have any clothes on  none of the above   
52152                                 would you eat this  none of the above   
52168                          will this clock keep time  none of the above   

       correct  partial  
51801        1      1.0  
51845        1      1.0  
51852        1      1.0  
51888        1      1.0  
51911        1      1.0  
51979        1      1.0  
52020        1      1.0  
52055        1      1.0  
52152        1      1.0  
52168        1      1.0

In [17]:
# examples of <60% accuracy for yes/no answer type with incorrect answers - 1555 data points
df[(df.answer_type=='yes/no') & \
   ((df.question_type=='are there any') | (df.question_type=='none of the above')) & \
   (df.correct==0)].tail(10)

annotations  answer_id  \
51066  [radio tv, radio and television, yes, radio an...   18692710   
51247    [yes, yes, yes, yes, yes, no, yes, no, yes, no]   27385900   
51428  [yes, yes, yes, yes, yes, yes, yes, yes, yes, ...   51220600   
51467    [yes, no, yes, yes, yes, yes, yes, no, no, yes]    3139020   
51547           [no, no, no, no, no, no, no, no, no, no]   40843920   
51828  [yes, yes, yes, yes, yes, yes, yes, yes, yes, ...   37031520   
51879           [no, no, no, no, no, no, no, no, no, no]   15277620   
51901          [no, no, no, no, no, no, no, no, yes, no]   29378200   
51985           [no, no, no, no, no, no, no, no, no, no]   53321720   
52038           [no, no, no, no, no, no, no, no, no, no]   34400520   

      answer_str answer_type complement_id  image_id  one_hot_index  \
51066        yes      yes/no          None    186927              1   
51247        yes      yes/no          None    273859              1   
51428        yes      yes/no          None    512206              1   
51467        yes      yes/no          None     31390              1   
51547         no      yes/no          None    408439              2   
51828        yes      yes/no          None    370315              1   
51879         no      yes/no          None    152776              2   
51901         no      yes/no          None    293782              2   
51985         no      yes/no          None    533217              2   
52038         no      yes/no          None    344005              2   

      predicted_answer  question_id  \
51066               no      1869271   
51247               no      2738590   
51428               no      5122060   
51467               no       313902   
51547              yes      4084392   
51828               no      3703152   
51879              yes      1527762   
51901              yes      2937820   
51985              yes      5332172   
52038              yes      3440052   

                                            question_str      question_type  \
51066                                anything on the bed  none of the above   
51247                    does near the door need painted  none of the above   
51428             are there any butterflies in the photo      are there any   
51467                        does everyone have on short  none of the above   
51547  if a person swan to shore would the be able to...  none of the above   
51828                               does he wear glasses  none of the above   
51879                 did these come from a money garden  none of the above   
51901             will this street sign wave in the wind  none of the above   
51985                     does it look like a cloudy day  none of the above   
52038                                  will it rain soon  none of the above   

       correct   partial  
51066        0  0.000000  
51247        0  1.000000  
51428        0  0.000000  
51467        0  1.000000  
51547        0  0.000000  
51828        0  0.000000  
51879        0  0.000000  
51901        0  0.333333  
51985        0  0.000000  
52038        0  0.000000

In [18]:
# how does model compare to humans?
# what percentage of incorrect answers predicted by model are also predicted by humans?
num_incorrect = df[(df.answer_type=='yes/no') & (df.correct==0)]['annotations'].count()
num_atleast1 = df[(df.answer_type=='yes/no') & (df.correct==0) & (df.partial>0)]['annotations'].count() 
num_atleast2 = df[(df.answer_type=='yes/no') & (df.correct==0) & (df.partial>0.35)]['annotations'].count() 
num_atleast3 = df[(df.answer_type=='yes/no') & (df.correct==0) & (df.partial==1)]['annotations'].count() 

print('Percentage of incorrect answers predicted by at least one human: {:.1%}'.format(num_atleast1/num_incorrect))
print('Percentage of incorrect answers predicted by at least two human: {:.1%}'.format(num_atleast2/num_incorrect))
print('Percentage of incorrect answers predicted by at least three human: {:.1%}'.format(num_atleast3/num_incorrect))

Percentage of incorrect answers predicted by at least one human: 48.7%
Percentage of incorrect answers predicted by at least two human: 30.3%
Percentage of incorrect answers predicted by at least three human: 19.5%


### 1b. Other Answer Type

In [19]:
# accuracy by question type
df[df.answer_type=='other'].groupby(['question_type'])['correct', 'partial'].mean().sort_values(['correct'])

correct   partial
question_type                               
are                       0.000000  0.111111
how many people are in    0.000000  0.000000
can you                   0.000000  0.266667
how many people are       0.000000  0.000000
how many                  0.172414  0.183908
are they                  0.230769  0.358974
where is the              0.282813  0.363542
why                       0.284848  0.305051
why is the                0.298701  0.359307
what color                0.303116  0.340888
how                       0.316667  0.366667
what is in the            0.333333  0.387636
is the woman              0.333333  0.370370
are there                 0.357143  0.428571
where are the             0.358650  0.431786
what is                   0.363072  0.417895
is that a                 0.368421  0.385965
what color is             0.371758  0.447646
who is                    0.373391  0.422031
what is on the            0.391924  0.431512
none of the above         0.395031  0.442650
is there                  0.428571  0.619048
which                     0.430147  0.496324
what color is the         0.434685  0.489342
what color are the        0.449541  0.505734
what does the             0.453125  0.491319
what time                 0.453416  0.631470
is this person            0.454545  0.454545
what is the color of the  0.459227  0.510730
does this                 0.461538  0.512821
what is the               0.465456  0.509005
what is the woman         0.467662  0.502488
what are the              0.471229  0.522032
is there a                0.500000  0.666667
what is the name          0.500000  0.500000
is he                     0.500000  0.500000
what                      0.509046  0.557408
are these                 0.518519  0.641975
is the                    0.520942  0.615183
is                        0.526316  0.605263
what are                  0.528239  0.570321
is the man                0.535714  0.595238
are the                   0.550562  0.595506
what is the person        0.554502  0.617694
is the person             0.555556  0.620370
what kind of              0.556027  0.616299
is this a                 0.570093  0.676012
what is the man           0.570481  0.622996
is this an                0.571429  0.571429
does the                  0.571429  0.702381
what is this              0.572626  0.618250
what type of              0.574813  0.637573
is this                   0.589928  0.657074
what brand                0.594595  0.648649
what animal is            0.674107  0.690476
is it                     0.753425  0.780822
was                       0.809524  0.809524
do                        0.857143  0.857143
what sport is             0.858934  0.881923
what room is              0.864407  0.878531
do you                    1.000000  1.000000

In [20]:
# examples of incorrect predictions for other answer type
df[(df.answer_type=='other') & (df.correct==0)].tail(10)

annotations  answer_id  \
52183  [lights, lights, lights, lights, lights, light...   53979120   
52186  [black and white, black and white striped, bla...   18924110   
52187  [white silver, gray and white, white, white an...   18924120   
52189  [green and yellow, green, yellow and green, gr...   11380110   
52193  [helmet, helmet, helmet, helmet, helmet, helme...   40976310   
52202  [flowers, flowers, flowers, sunglass, flowers ...    8603610   
52205  [log, tree, tree, tree, log, tree, log, tree, ...   53298910   
52207  [yellow, yellow, yellow white black, yellow, y...   20132610   
52208  [red black, red white black, red, orange, red ...   20132620   
52212  [table, on table, on right, on right, by napki...    1603000   

             answer_str answer_type complement_id  image_id  one_hot_index  \
52183            lights       other          None    539791            321   
52186             black       other          None    189241             11   
52187             white       other          None    189241              5   
52189  green and yellow       other          None    113801            814   
52193            helmet       other          None    409763             91   
52202           flowers       other          None     86036             73   
52205              tree       other          None    532989            110   
52207            yellow       other          None    201326             12   
52208     red and black       other          None    201326            436   
52212             table       other          None     16030             77   

      predicted_answer  question_id  \
52183            birds      5397912   
52186              red      1892411   
52187              red      1892412   
52189              red      1138011   
52193              hat      4097631   
52202              hat       860361   
52205            rocks      5329891   
52207            white      2013261   
52208              red      2013262   
52212         on table       160300   

                                          question_str      question_type  \
52183                   what's at the top of the poles  none of the above   
52186            what color apron is the woman wearing         what color   
52187                       what color is the vehicles  what color is the   
52189                            what color is the bus  what color is the   
52193                        what is on the man's head     what is on the   
52202                       what is on the girl's head     what is on the   
52205  what is laying on the ground behind the giraffe            what is   
52207                         what color is he wearing      what color is   
52208                        what color is the racquet  what color is the   
52212                                where is the fork       where is the   

       correct   partial  
52183        0  0.000000  
52186        0  0.000000  
52187        0  0.000000  
52189        0  0.000000  
52193        0  0.333333  
52202        0  0.000000  
52205        0  0.000000  
52207        0  0.000000  
52208        0  0.333333  
52212        0  1.000000

In [21]:
# create column with number of words in answers
df['answer_length'] = df.apply(lambda data: len(data['answer_str'].split()), axis=1)
df.tail(10)

annotations  answer_id  \
52203  [yes, yes, yes, yes, yes, yes, yes, yes, yes, ...    8603620   
52204  [yes, yes, yes, yes, yes, yes, yes, yes, yes, ...   53298900   
52205  [log, tree, tree, tree, log, tree, log, tree, ...   53298910   
52206  [yes, yes, yes, yes, yes, yes, yes, no, yes, yes]   20132600   
52207  [yellow, yellow, yellow white black, yellow, y...   20132610   
52208  [red black, red white black, red, orange, red ...   20132620   
52209  [night, night, night, night, night, night, nig...   47747000   
52210   [yes, yes, yes, no, no, yes, yes, yes, yes, yes]   47747010   
52211  [stop, stop, stop, stop, stop, stop, stop, sto...   47747020   
52212  [table, on table, on right, on right, by napki...    1603000   

          answer_str answer_type complement_id  image_id  one_hot_index  \
52203            yes      yes/no          None     86036              1   
52204            yes      yes/no          None    532989              1   
52205           tree       other          None    532989            110   
52206            yes      yes/no          None    201326              1   
52207         yellow       other          None    201326             12   
52208  red and black       other          None    201326            436   
52209          night       other          None    477470             92   
52210            yes      yes/no          None    477470              1   
52211           stop       other          None    477470             50   
52212          table       other          None     16030             77   

      predicted_answer  question_id  \
52203               no       860362   
52204               no      5329890   
52205            rocks      5329891   
52206              yes      2013260   
52207            white      2013261   
52208              red      2013262   
52209            night      4774700   
52210              yes      4774701   
52211             stop      4774702   
52212         on table       160300   

                                          question_str      question_type  \
52203                            are the girls topless            are the   
52204                        is the giraffe at the zoo             is the   
52205  what is laying on the ground behind the giraffe            what is   
52206      is the tennis player wearing a nike t shirt             is the   
52207                         what color is he wearing      what color is   
52208                        what color is the racquet  what color is the   
52209            what time of day was this photo taken          what time   
52210         is this the right atmosphere for dracula            is this   
52211            what does the traffic light say to do      what does the   
52212                                where is the fork       where is the   

       correct   partial  answer_length  
52203        0  0.000000              1  
52204        0  0.000000              1  
52205        0  0.000000              1  
52206        1  1.000000              1  
52207        0  0.000000              1  
52208        0  0.333333              3  
52209        1  1.000000              1  
52210        1  1.000000              1  
52211        1  1.000000              1  
52212        0  1.000000              1

In [22]:
# number of incorrect predictions by answer length
df[(df.answer_type=='other') & (df.correct==0)].groupby(['answer_length'])['annotations'].count()

answer_length
0        5
1    11058
2      692
3      304
4        8
Name: annotations, dtype: int64

In [23]:
# number of correct predictions by answer length
df[(df.answer_type=='other') & (df.correct==1)].groupby(['answer_length'])['annotations'].count()

answer_length
1    9940
2     529
3     147
4       3
Name: annotations, dtype: int64

### 1c. Number Answer Type

In [24]:
# accuracy by question type
df[df.answer_type=='number'].groupby(['question_type'])['correct', 'partial'].mean().sort_values(['correct'])

correct   partial
question_type                             
are there               0.000000  0.000000
what are the            0.000000  0.000000
was                     0.000000  0.333333
is this person          0.000000  0.000000
what type of            0.000000  0.000000
is there                0.000000  1.000000
is the man              0.000000  0.000000
is the                  0.000000  0.000000
is this                 0.000000  0.000000
is he                   0.000000  0.000000
what is the             0.048193  0.080321
what number is          0.050847  0.050847
which                   0.076923  0.179487
what                    0.093023  0.162791
none of the above       0.146341  0.276423
how                     0.220588  0.245098
how many people are in  0.244344  0.313725
how many people are     0.291339  0.372703
how many                0.293448  0.385208
what is                 0.400000  0.466667
is                      0.500000  0.833333
does the                0.500000  0.500000
what time               0.500000  0.444444
what does the           0.625000  0.625000

In [25]:
# number of data points by question type
df[(df.answer_type=='number')].groupby(['question_type'])['annotations'].count().sort_values(ascending=False)

question_type
how many                  5449
how many people are        508
how many people are in     221
how                        136
what                       129
what number is             118
what is the                 83
none of the above           41
which                       13
what does the                8
what time                    6
what is                      5
does the                     4
is                           2
is this                      2
what are the                 2
is the                       1
is he                        1
is the man                   1
is there                     1
what type of                 1
is this person               1
was                          1
are there                    1
Name: annotations, dtype: int64

In [26]:
# examples of incorrect predictions for other answer type
# df[(df.answer_type=='number') & (df.correct==1) & (df.question_type=='how many people are')].tail(10)
df[(df.answer_type=='number') & (df.correct==0)].tail(10)

annotations  answer_id  \
52103  [more than 15, 20, many, many, 60, 20, 10, 22,...   31425110   
52106                  [2, fork, 3, 4, 2, 1, 2, 2, 3, 2]   53063020   
52136                   [4, 7, 4, 5, 2, 3, 6, 6, 10, 15]   22553200   
52137           [20, 20, 20, 20, 20, 20, 20, 20, 60, 20]   22553210   
52149                 [1, 1, 1, 1, 1, obits, 1, 1, 1, 1]   51762910   
52162                    [5, 1, 6, 4, 4, 3, 4, 4, 10, 4]   23307910   
52173              [20, 4, 20, 20, 9, 3, lot, 1, 23, 19]   16313220   
52180                     [2, 2, 2, 2, 2, 2, 2, 2, 2, 2]    3766000   
52188                     [0, 0, 0, 0, 0, 1, 0, 0, 0, 0]   11380100   
52191                     [2, 2, 2, 2, 2, 2, 2, 2, 2, 2]   40976320   

      answer_str answer_type complement_id  image_id  one_hot_index  \
52103       many      number          None    314251            161   
52106          2      number          None    530630              3   
52136          4      number          None    225532              9   
52137         20      number          None    225532            112   
52149          1      number          None    517629              4   
52162          4      number          None    233079              9   
52173         20      number          None    163132            112   
52180          2      number          None     37660              3   
52188          0      number          None    113801             19   
52191          2      number          None    409763              3   

      predicted_answer  question_id  \
52103                1      3142511   
52106                4      5306302   
52136                2      2255320   
52137                2      2255321   
52149                2      5176291   
52162                2      2330791   
52173                3      1631322   
52180                3       376600   
52188                2      1138010   
52191               13      4097632   

                                            question_str question_type  \
52103                     how many trees are by the road      how many   
52106                how many silverware items are there      how many   
52136             how many buildings are in this picture      how many   
52137                                       how many mph      how many   
52149                     how many doors are in the room      how many   
52162                  how many benches are in the lobby      how many   
52173                how many lights are below the plane      how many   
52180                     how many items are in the hand      how many   
52188                    how many boats are in the photo      how many   
52191  what is the number on the back of the batter o...   what is the   

       correct   partial  answer_length  
52103        0  0.000000              1  
52106        0  0.333333              1  
52136        0  0.333333              1  
52137        0  0.000000              1  
52149        0  0.000000              1  
52162        0  0.000000              1  
52173        0  0.333333              1  
52180        0  0.000000              1  
52188        0  0.000000              1  
52191        0  0.000000              1

In [27]:
# create column with unk token flag
df['unk_flag'] = df.apply(lambda data: data['predicted_answer']=='<unk>', axis=1)
df.head(5)

annotations  answer_id answer_str  \
0           [16, 16, 16, 16, 16, 16, 16, 16, 16, 16]    9786500         16   
1  [yes, yes, yes, yes, yes, yes, yes, yes, yes, ...    9786510        yes   
2  [container, frisbee golf, frisbee golf goal, f...    9786520    frisbee   
3      [garbage, no, no, no, no, no, no, no, no, no]   57484500         no   
4  [yes, yes, yes, yes, yes, yes, yes, yes, yes, ...   57484510        yes   

  answer_type complement_id  image_id  one_hot_index predicted_answer  \
0      number          None     97865            241     fire hydrant   
1      yes/no          None     97865              1               no   
2       other          None     97865             24     fire hydrant   
3      yes/no          None    574845              2               no   
4      yes/no          None    574845              1              yes   

   question_id                     question_str question_type  correct  \
0       978650                     what # is it          what        0   
1       978651             is there people here      is there        0   
2       978652  what is the object on the right   what is the        0   
3      5748450                   is this inside       is this        1   
4      5748451         could someone sleep here         could        1   

   partial  answer_length  unk_flag  
0      0.0              1     False  
1      0.0              1     False  
2      0.0              1     False  
3      1.0              1     False  
4      1.0              1     False

In [28]:
# number of incorrect answers by unk token flag
df[(df.answer_type=='number') & (df.correct==0)].groupby(['unk_flag'])['annotations'].count()

unk_flag
False    4862
Name: annotations, dtype: int64

In [29]:
# examples of incorrect predictions that are not unk tokens
df[(df.answer_type=='number') & (df.correct==0) & (df.predicted_answer == '<unk>')].tail(50)

Empty DataFrame
Columns: [annotations, answer_id, answer_str, answer_type, complement_id, image_id, one_hot_index, predicted_answer, question_id, question_str, question_type, correct, partial, answer_length, unk_flag]
Index: []

In [30]:
# examples of incorrect predictions that are not unk tokens
df[(df.answer_type=='number') & (df.correct==0) & (df.predicted_answer != '<unk>')].head(10)

annotations  answer_id answer_str  \
0      [16, 16, 16, 16, 16, 16, 16, 16, 16, 16]    9786500         16   
10               [2, 2, 2, 2, 2, 2, 2, 2, 2, 2]   19991810          2   
46    [16, 15, 14, 13, 18, 10, 16, 15, lot, 13]   38084200         13   
50               [5, 4, 4, 5, 5, 5, 5, 5, 5, 5]   57433200          5   
81               [2, 2, 2, 2, 2, 2, 2, 4, 2, 2]   44040010          2   
83               [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]   53446720          1   
97               [1, 1, 1, 1, 1, 2, 1, 1, 1, 1]   30653620          1   
134  [bananas, 7, boggles, 7, 7, 7, 7, 7, 6, 7]   49597510          7   
135              [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]   18817300          0   
136              [5, 5, 5, 5, 5, 5, 5, 5, 5, 5]   18817320          5   

    answer_type complement_id  image_id  one_hot_index predicted_answer  \
0        number          None     97865            241     fire hydrant   
10       number          None    199918              3                1   
46       number          None    380842            137                3   
50       number          None    574332             14                2   
81       number          None    440400              3                1   
83       number          None    534467              4                2   
97       number          None    306536              4                2   
134      number          None    495975             31                3   
135      number          None    188173             19                2   
136      number          None    188173             14                2   

     question_id                                      question_str  \
0         978650                                      what # is it   
10       1999181         how many waterfalls are entering the pool   
46       3808420  how many crosswalk stripes painted on the street   
50       5743320                       how many umbrellas are open   
81       4404001              how many people are behind the woman   
83       5344672                  how many towels are in the photo   
97       3065362                                   how many clocks   
134      4959751                how many bunches are on this scene   
135      1881730                      how many kangaroos are there   
136      1881732                    how many cars are in the photo   

           question_type  correct   partial  answer_length  unk_flag  
0                   what        0  0.000000              1     False  
10              how many        0  0.000000              1     False  
46              how many        0  0.000000              1     False  
50              how many        0  0.000000              1     False  
81   how many people are        0  0.000000              1     False  
83              how many        0  0.000000              1     False  
97              how many        0  0.333333              1     False  
134             how many        0  0.000000              1     False  
135             how many        0  0.000000              1     False  
136             how many        0  0.000000              1     False

#### Summary
- Best accuracy (66%): Yes/No answer type
- Second accuracy (28%): Other answer type
- Worst accuracy (22%): Number answer type

**Yes/No Answer Type**  
The model does quite well consistently for all question types with accuracy above 50% except for three question types including "why", "what", and "what is the".  As we can see, the phrasing of these question types do not point to a yes or no answer so it's not a surprise that the model does poorly on these questions.  For other question types, the model does the best when the questions have clear and direct answers and it performs poorly when the questions are abstract, subjective, or require common sense knowledge.  It's also interesting to note that for the incorrect predictions almost half of them (47%) were also predicted by at least one human and almost 29% were predicted by at least two human.

**Other Answer Type**
We explored whether multi-word answers led to low accuracy for this answer type since multi-word phrases are more likely to be excluded from the training vocabulary.  However, we found that most of the incorrect predictions (80%) are made up of one-word answer and only 20% are made up of multi-word answers.  In general, the model does the best in predicting rooms, animals, sports, and colors for this type of question.


**Number Answer Type**
The model does the worst for number answer type with only 22% accuracy.  Most of the questions for this answer type involve counting i.e. "how many".  Around 20% of the incorrect predictions have the UNK token as the predicted answers.  We found that a lot of the UNK token predictions involve answers that relate to time or number sequences (such as bus number or number on a jersey) that need to be identified on objects.   



For future iterations we can try to measure accuracy by measuring distance between word vectors so that words / phrases with similar semantic meanings can be given credit (ex: nighttime vs. night).  We can also find better ways to tokenize time and numbers and explore different UNK replacement techniques.
